Byte updates using bit fields require extension to full bytes

tautschnig opened the pull request:
A byte update with an update value that has a size that's not a multiple
of bytes must not overwrite bits that are not in the update value. We
previously generated extractbits expressions that would go out of
bounds, which tripped up the SMT solvers. No longer generating these
out-of-bounds expressions makes the SMT solvers happy on the Promotion3
test, which exercised this feature.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs commented on src/goto-programs/builtin_functions.cpp:
```diff
@@ -699,7 +700,7 @@ void goto_convertt::do_havoc_slice(
     "assertion havoc_slice " + from_expr(ns, identifier, ok_expr));
 
-  const array_typet array_type(char_type(), arguments[1]);
+  const array_typet array_type(char_type(), simplify_expr(arguments[1], ns));
```
I get that this might be the right thing to do and that it fixes a bug but this doesn't immediately seem linked to the description of the PR.

tautschnig replied:
Thanks, addressed in two ways: 1) I've moved this change to a commit of its own; 2) I've added a comment explaining when a `DATA_INVARIANT` newly introduced here could fail (which really was what made me detect the above).

martin-cs replied:
Amazing!  Thank you.

martin-cs reviewed (approved):
Looks as expected.  Fewer tests marked as buggy is always good.

tautschnig committed:
Simplify away type cast

The array size may include a type cast, possibly of a constant.
Therefore, the size would be found to be constant via simplification,
but other ways of computing the size may fail to derive a constant.

tautschnig committed:
Byte updates using bit fields require extension to full bytes

A byte update with an update value that has a size that's not a multiple
of bytes must not overwrite bits that are not in the update value. We
previously generated extractbits expressions that would go out of
bounds, which tripped up the SMT solvers. No longer generating these
out-of-bounds expressions makes the SMT solvers happy on the Promotion3
test, which exercised this feature.
